server/pulp/server/webservices/views/consumers.py

```diff
@@ -0,0 +1,1082 @@
+
+        link = add_link(consumer)
+
+        document = {
+            'consumer': consumer,
+            'certificate': certificate
+        }
+        response = generate_json_response_with_pulp_encoder(document)
+        redirect_response = generate_redirect_response(response, link['_href'])
+        return redirect_response
+
+
+class ConsumerResourceView(View):
+    """
+    View for single consumer.
+    """
+
+    @auth_required(authorization.READ)
+    def get(self, request, consumer_id):
+        """
+        Return a serialized object representing the requested consumer.
+
+        :param request: WSGI request object
+        :type request: django.core.handlers.wsgi.WSGIRequest
+        :param consumer_id: id for the requested consumer
+        :type consumer_id: str
+
+        :return: Response containing data for the requested consumer
+        :rtype: django.http.HttpResponse
+        """
+
+        query_params = request.GET
+        manager = factory.consumer_manager()
+        consumer = manager.get_consumer(consumer_id)
+        consumer = expand_consumers(query_params, [consumer])[0]
+        add_link(consumer)
+        return generate_json_response_with_pulp_encoder(consumer)
+
+    @auth_required(authorization.DELETE)
+    def delete(self, request, consumer_id):
+        """
+        Delete a specified consumer.
+
+        :param request: WSGI request object
+        :type request: django.core.handlers.wsgi.WSGIRequest
+        :param consumer_id: id for the requested consumer
+        :type consumer_id: str
+
+        :return: An empty response
+        :rtype: django.http.HttpResponse
+        """
+
+        manager = factory.consumer_manager()
+        response = manager.unregister(consumer_id)
+        return generate_json_response(response)
+
+    @auth_required(authorization.UPDATE)
+    @json_body_required
+    def put(self, request, consumer_id):
+        """
+        Update a specified consumer.
+
+        :param request: WSGI request object
+        :type request: django.core.handlers.wsgi.WSGIRequest
+        :param consumer_id: id for the requested consumer
+        :type consumer_id: str
+
+        :return: Response representing the updated consumer
+        :rtype: django.http.HttpResponse
+        """
+
+        params = request.body_as_json
+        delta = params.get('delta')
+        manager = factory.consumer_manager()
+        consumer = manager.update(consumer_id, delta)
+        add_link(consumer)
+        return generate_json_response_with_pulp_encoder(consumer)
+
+
+class ConsumerBindingsView(View):
+    """
+    View for Consumer bindings - represents the collection of
+    objects used to associate a consumer and a repo-distributor
+    association.  Users wanting to create this association will
+    create an object in this collection.  Both bind and unbind
+    is idempotent.
+    """
+
+    @auth_required(authorization.READ)
+    def get(self, request, consumer_id, repo_id=None):
```

Comment:
There actually is a problem with this pattern. Having 2 urls that point to this View with different numbers of arguments is effectively overloading this View. The problem is that `/v2/consumers/<consumer_id>/bindings/` accepts only GET and POST, while `/v2/consumers/<consumer_id>/bindings/<repo_id>/` should only accept GET but does accept POST requests.

So if you do a POST request to the latter, it will return a 500:

`curl -i -H "Accept: application/json" -H "WebFrameworkSwitch: django" -X POST -d '{}' -k -u admin:admin  "https://localhost/pulp/api/v2/consumers/c1/bindings/zoo/"`

    HTTP/1.1 500 INTERNAL SERVER ERROR
    Date: Thu, 26 Mar 2015 17:56:25 GMT
    Server: Apache/2.4.10 (Fedora) OpenSSL/1.0.1k-fips mod_wsgi/4.3.2 Python/2.7.8
    Content-Encoding: utf-8
    Content-Length: 1267
    Connection: close
    Content-Type: application/json

    {"http_request_method": "POST", "exception": ["TypeError: post() got an unexpected keyword argument 'repo_id'\n"], "error_message": "post() got an unexpected keyword argument 'repo_id'", "_href": "/pulp/api/v2/consumers/c1/bindings/zoo/", "http_status": 500, "traceback": ["  File \"/usr/lib/python2.7/site-packages/django/core/handlers/base.py\", line 112, in get_response\n    response = wrapped_callback(request, *callback_args, **callback_kwargs)\n", "  File \"/usr/lib/python2.7/site-packages/django/views/generic/base.py\", line 69, in view\n    return self.dispatch(request, *args, **kwargs)\n", "  File \"/usr/lib/python2.7/site-packages/django/views/generic/base.py\", line 87, in dispatch\n    return handler(request, *args, **kwargs)\n", "  File \"/home/austin/devel/pulp/server/pulp/server/webservices/controllers/decorators.py\", line 241, in _auth_decorator\n    return _verify_auth(self, operation, super_user_only, method, *args, **kwargs)\n", "  File \"/home/austin/devel/pulp/server/pulp/server/webservices/controllers/decorators.py\", line 195, in _verify_auth\n    value = method(self, *args, **kwargs)\n", "  File \"/home/austin/devel/pulp/server/pulp/server/webservices/views/util.py\", line 90, in wrapper\n    return func(*args, **kwargs)\n"]}

The simple solution would be set a default value for repo_id like `repo_id=None` on POST. Then you can do this:

    if repo_id:
        raise HttpResponseNotFound()